### PR TITLE
Extend mathematic operations support in the kmath-ast parser 

### LINKED ISSUE #117: Support all mathematic operations in the kmath-ast parser
add support for singular, unary and prefix binary operations.

### PR DESCRIPTION
**Pull request by CommanderTvis**
- Add reference ANTLR grammar.
- Add symbols support in ASM MST matcher.
- Extend parser to support singular values (like i in ComplexField) as well as unary and binary functions.

closes #117

### CODE REVIEW

**Comment by CommanderTvis** [comment]
closes #117

**Comment by CommanderTvis** [comment]
Converting it to draft because of better-parse bug.

**Comment by CommanderTvis** [comment]
h0tk3y/better-parse/issues#28

File: kmath-ast/reference/ArithmeticsEvaluator.g4 [hunk 1]
```diff
@@ -0,0 +1,59 @@
+grammar ArithmeticsEvaluator;
+
+fragment DIGIT: '0'..'9';
+fragment LETTER: 'a'..'z';
```
Comment (Shimuuar) [inline kmath-ast/reference/ArithmeticsEvaluator.g4:4]: What about uppercase letters?

File: kmath-ast/reference/ArithmeticsEvaluator.g4 [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+fragment UNDERSCORE: '_';
+
+ID: (LETTER | UNDERSCORE) (LETTER | UNDERSCORE | DIGIT)*;
+NUM: (DIGIT | '.')+ ([eE] MINUS? DIGIT+)?;
```
Comment (Shimuuar) [inline kmath-ast/reference/ArithmeticsEvaluator.g4:8]: `1.23e+3` generally + in exponent is generally allowed

File: kmath-ast/reference/ArithmeticsEvaluator.g4 [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+powChain
+    : term (POW term)*
+    ;
+
+
+divMulChain
+    : powChain ((PLUS | MINUS) powChain)*
+    ;
+
+subSumChain
```
Comment (Shimuuar) [inline kmath-ast/reference/ArithmeticsEvaluator.g4:53]: Is operator precedence right? Looks suspicious. It would be nice to add test for it

Comment (CommanderTvis) [reply kmath-ast/reference/ArithmeticsEvaluator.g4:53]: Tests were added

File: kmath-ast/reference/ArithmeticsEvaluator.g4 (hunk 2 above)
Comment (CommanderTvis) [reply kmath-ast/reference/ArithmeticsEvaluator.g4]: Fixed

File: kmath-ast/reference/ArithmeticsEvaluator.g4 (hunk 1 above)
Comment (CommanderTvis) [reply kmath-ast/reference/ArithmeticsEvaluator.g4:4]: Added.

**Pull request merged**